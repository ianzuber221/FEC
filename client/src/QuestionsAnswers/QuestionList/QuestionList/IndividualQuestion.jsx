import React, { useState, useEffect } from 'react';
import styled from 'styled-components';
import _ from 'underscore';
import { getAnswers, postAnswer } from '../../lib/api/githubAPI';
import getImageUrl from '../../lib/api/cloudinaryAPI';
import Helpful from '../../components/Helpful';
import Report from '../../components/Report';
import AddAnswerForm from '../AddAnswer/AddAnswerForm';
import AnswerList from '../AnswerList/AnswerList';

function IndividualQuestion({ productName, question, renderQuestions, keyword }) {
  const [showModal, setShowModal] = useState(false);
  const [answerList, setAnswerList] = useState(sortAnswers(question.answers));
  const [page, setPage] = useState(1);
  const [count, setCount] = useState(100);
  const [helpfulCount, setHelpfulCount] = useState(question.question_helpfulness);

  const renderAnswers = () => {
    getAnswers(question.question_id, page, count)
      .then((result) => {
        if (result.data.results.length === 0) {
          return;
        }
        setAnswerList(sortAnswers(result.data.results));
      });
  };

  const addAnswer = (formValues) => {
    if (formValues.photos) {
      const photoUrlPromise = formValues.photos.map((photo) => getImageUrl(photo.file))
      Promise.all(photoUrlPromise)
        .then((result) => {
          formValues.photos = result;
          postAnswer(question.question_id, formValues)
            .then(() => {
              renderAnswers();
            });
        });
    } else {
      postAnswer(question.question_id, formValues)
        .then(() => {
          renderAnswers();
        });
    }
  };

  useEffect(() => {
    if (answerList.length >= count) {
      setCount(count * 2);
    }
  }, [answerList]);

  return (
    <DivContainer show={question.question_body.match(new RegExp(keyword, 'i'))}>
      <DivQuestion>
        <QContainer>
          <Title>Q:</Title>
          <BoldBody>{question.question_body}</BoldBody>
        </QContainer>
        <OptionsDiv>
          <Helpful
            id={question.question_id}
            type="question"
            currentCount={helpfulCount}
            setHelpfulCount={setHelpfulCount}
            renderComponent={renderQuestions}
            tabIndex="0"
          />
          <PaddedSpan>{' | '}</PaddedSpan>
          <Report
            id={question.question_id}
            type="question"
          />
          <PaddedSpan>{' | '}</PaddedSpan>
          <ClickableU onClick={() => setShowModal(true)} onKeyDown={() => setShowModal(true)} role="button" tabIndex="-1">
            Add Answer
          </ClickableU>
        </OptionsDiv>
      </DivQuestion>
      <AnswerList
        answerList={answerList}
        renderAnswers={renderAnswers}
        Title={Title}
      />
      <AddAnswerForm
        question={question}
        submitHandler={addAnswer}
        productName={productName}
        show={showModal}
        setShowModal={setShowModal}
      />
    </DivContainer>
  );
}

export default IndividualQuestion;

const sortAnswers = (answers) => (
  _(answers).chain()
    .sortBy('helpfulness')
    .sortBy((answer) => answer.answerer_name.toLowerCase() === 'seller')
    .reverse()
    .value()
);

const DivContainer = styled.div`
  display: ${props => props.show ? 'block' : 'none'}
`;

const DivQuestion = styled.div`
  display: flex;
  justify-content: space-between;
  align-items: center;
  margin: 15px 0;
  padding-right: 10px;
`;

const Title = styled.span`
  width: 30px;
  font-weight: bold;
`;

const QContainer = styled.div`
  display: flex;
  width: 60%;
`;

const BoldBody = styled.b`
  width: 100%;
`;

const OptionsDiv = styled.div`
  font-size: 11px;
  font-weight: 300;
  color: #77787a;
  display: flex;
  max-width: 40%;
  align-items: center;
`;

const PaddedSpan = styled.div`
  padding: 0 10px;
  font-size: 15px;
`;

const ClickableU = styled.u`
  &:hover {
    cursor: pointer;
    opacity: 60%;
  }
`;