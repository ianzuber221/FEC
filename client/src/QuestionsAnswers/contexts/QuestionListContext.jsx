import React, { useState, useContext } from 'react';

const QuestionListContext = React.createContext();
const UpdateQuestionListContext = React.createContext();

export function useQuestionList() {
  return useContext(QuestionListContext);
}

export function useUpdateQuestionList() {
  return useContext(UpdateQuestionListContext);
}

export function QuestionListProvider({ children }) {
  const [questionList, setQuestionList] = useState([]);

  function updateQuestionList(newQuestionList) {
    setQuestionList(newQuestionList);
  }

  return (
    <QuestionListContext.Provider value={questionList}>
      <UpdateQuestionListContext.Provider value={updateQuestionList}>
        { children }
      </UpdateQuestionListContext.Provider>
    </QuestionListContext.Provider>
  );
}
