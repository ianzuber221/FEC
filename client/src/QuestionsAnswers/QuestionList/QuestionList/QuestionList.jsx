import React, { useRef } from 'react';
import styled from 'styled-components';
import IndividualQuestion from './IndividualQuestion';
import { useQuestionList } from '../../contexts/QuestionListContext';

function QuestionList({ renderQuestions, keyword, productName, expanded, page, setPage, hasMore }) {
  const questions = useQuestionList();
  const listInnerRef = useRef();

  if (questions.length === 0) {
    return (
      <NoDataDiv id="question-list">
        No Questions Available.
      </NoDataDiv>
    );
  }

  // let filteredQuestions = questions
  //   .filter((question) => (
  //     question.question_body.match(new RegExp(keyword, 'i'))
  //   ));

  const onScroll = async () => {
    if (listInnerRef.current) {
      const { offsetHeight, scrollTop, scrollHeight } = listInnerRef.current;
      if (scrollTop + offsetHeight + 10 >= scrollHeight) {
        if (hasMore) {
          setPage(page + 1);
          renderQuestions();
        }
      }
    }
  };

  return (
    <div id="question-list">
      {expanded
      && (
        <Scroller onScroll={onScroll} ref={listInnerRef} id="question-scroller">
          {questions
            .map((question) => (
              <IndividualQuestion
                key={question.question_id}
                question={question}
                renderQuestions={renderQuestions}
                productName={productName}
                keyword={keyword}
              />
            ))}
        </Scroller>
      )}
      {!expanded
      && (
        <div>
          {questions
            .slice(0, 2)
            .map((question) => (
              <IndividualQuestion
                key={question.question_id}
                question={question}
                renderQuestions={renderQuestions}
                productName={productName}
                keyword={keyword}
              />
            ))}
        </div>
      )}

    </div>
  );
}

export default QuestionList;

const Scroller = styled.div`
  width: 100%;
  overflow-y: auto;
  max-height: 70vh;
`;

const NoDataDiv = styled.div`
  padding: 10px 0;
  font-size: 18px;
`;
