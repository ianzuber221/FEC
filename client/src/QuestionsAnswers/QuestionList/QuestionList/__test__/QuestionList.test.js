/**
 * @jest-environment jsdom
 */
/* eslint-disable react/jsx-filename-extension */

import React from 'react';
import { render } from '@testing-library/react';
import { jest } from '@jest/globals';
import QuestionList from '../QuestionList';
import '@testing-library/jest-dom/extend-expect';
import { QuestionListProvider } from '../../../contexts/QuestionListContext';

const renderQuestionMock = jest.fn();

it('should match snapshot', async () => {
  const { asFragment } = await render(
    <QuestionListProvider>
      <QuestionList
        renderQuestions={renderQuestionMock}
        keyword="sample"
        productName="sample"
        expanded
      />
    </QuestionListProvider>
  );

  // await waitForElementToBeRemoved(screen.getByText('No Questions Available.'));

  expect(asFragment()).toMatchSnapshot();
});

// it('should render after form is submitted', async () => {
//   const { getByTestId } = render (
//     <QuestionListProvider>
//       <QuestionList
//         productName="sample"
//         question={{
//           question_id: 543290,
//           question_helpfulness: 3,
//           question_body: 'new question',
//           answers: [],
//         }}
//         renderQuestions={renderMock}
//       />
//     </QuestionListProvider>
//   );


// });
