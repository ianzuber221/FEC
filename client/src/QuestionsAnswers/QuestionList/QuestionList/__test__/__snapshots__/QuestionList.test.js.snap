// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should match snapshot 1`] = `
<DocumentFragment>
  <div
    class="sc-kLLXSd ggZKBR"
    id="question-list"
  >
    No Questions Available.
  </div>
</DocumentFragment>
`;
